### PR TITLE
modules/github-runner: Improve description of `url`

### PR DESCRIPTION
**Pull request by Profpatsch**
We spent a whole afternoon debugging this, because upstream has very
bad software quality and the error messages were incredibly
misleading.

So let’s document it for the sanity of other people.

Btw, I think the implementation of our module is pretty brittle,
especially the part about diffing tokens to check whether they
changed. We should rather just request a new builder registration
every time, it’s not that much overhead, and always set `replace` so
it is idempotent.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by veehaitch — APPROVED** [review]
Fine with me. Best advice is IMHO to use the URL GitHub displays on the page to add a new runner in the "Configure" section (see the `--url` parameter):

image

> Btw, I think the implementation of our module is pretty brittle,
especially the part about diffing tokens to check whether they
changed. We should rather just request a new builder registration
every time, it’s not that much overhead, and always set replace so
it is idempotent.

I agree that the module is too complex and, hence, hard to follow. You cannot, however, do a registration on each start as the `token` is rather short-lived (~ 1 hour). To use this approach, we'd need support for PATs. It still wouldn't work for the registration tokens though.

**Comment by Profpatsch** [comment]
> You cannot, however, do a registration on each start as the `token` is rather short-lived (~ 1 hour). To use this approach, we'd need support for PATs. It still wouldn't work for the registration tokens though.

Hm, that’s a good point.

PAT support has been added to the runner end of last year iirc, but you are right that wouldn’t help much for the organization tokens.

Most of this debugging was that the github-runner codebase is truly not very good (I had to read a nontrivial amount of it and work with their horrible logging scheme to figure out why the NotFound error was thrown), and that apparently the (completely undocumented) registration endpoint changed behavior in the meantime.

**Comment by veehaitch** [comment]
> PAT support has been added to the runner end of last year iirc, but you are right that wouldn’t help much for the organization tokens.

That shouldn't be a problem: using an appropriately scoped PAT, you can query a registration token on a per repository level and also on an organization level. Still, there are good reasons you might not want to deploy a powerful PAT on a machine executing the GitHub Actions runner. Therefore, we should keep supporting the expiring registration tokens. Nevertheless, refactoring/rewriting the current module as well as introducing support for PATs is a good and appreciated idea.

> Most of this debugging was that the github-runner codebase is truly not very good (I had to read a nontrivial amount of it and work with their horrible logging scheme to figure out why the NotFound error was thrown), and that apparently the (completely undocumented) registration endpoint changed behavior in the meantime.

What should I say—you're just right here.
